### ISSUE TITLE
Add reflection support for individual ubo and ssbo members

### ISSUE THREAD

**Issue by soconne** [GPUOpen-LibrariesAndSDKs/V-EZ#24]
Original request made in #22

**Comment by SilvanVR** [comment]
Encountered another issue related to this.
Struct members are not parsed. Example:
layout (set = 2, binding = 0) uniform LIGHTS
{
	Light lights[MAX_LIGHTS]; // Assume struct Light exist
	int count; // 4 bytes
} _Lights;

In the VezPipelineResource the size of this buffer is just 4.
It would be nice to parse struct members aswell recursively.

**Comment by SilvanVR** [comment]
Another bug. The size of a "mat4" is reported as 16 bytes, but should be 64. Example:
layout (set = 3, binding = 0) uniform TEST
{
	mat4 view;
};
// Size reported as 16

**Comment by soconne** [comment]
Unfortunately SPIRV-Cross does not allow for reflecting on members of a UBOs and SSBOs in the same way as resource bindings.  The exposed interfaces, at least from what I've seen, do not return the individual names or byte offsets of members.  All that can be retrieved is information about the types themselves, independent of their position within the relevant struct.

As for the bugs you reported, I will fix these.

**Comment by SilvanVR** [comment]
I think you are misinformed. VEZ is using spirv-cross which allows all of that. I did it myself already two years ago. I will just post my function for parsing all members recursively, even structs in structs. A word of warning though, i did this with **very** little C++ experience. The code is really ugly:

```
    // Information about every data-member in a struct in a UBO
    struct BufferRange
    {
        std::string name;
        uint32_t    offset;
        uint32_t    range;
        uint32_t    index;
        DataType    dataType;
    };

    // Parse a UBO and return a list of names, offsets and size of the individual variables
    std::vector<BufferRange> parseUniformStruct(const spirv_cross::Compiler& comp, const spirv_cross::SPIRType& spirType,
                                                uint32_t parentOffset, const std::string& parentName)
    {
        std::vector<BufferRange> memberRanges;

        for (unsigned int i = 0; i < spirType.member_types.size(); i++)
        {
            // Get the member-name and add the "parent"-name infront of it e.g. "baseLight-" + color
            const std::string& memberName = parentName + comp.get_member_name(spirType.self, i);
            // Get the offset of the member and add the parent-offset to it
            uint32_t offset = parentOffset + comp.get_member_decoration(spirType.self, i, spv::DecorationOffset);

            const spirv_cross::SPIRType& memberType = comp.get_type(spirType.member_types[i]);

            if (memberType.basetype == spirv_cross::SPIRType::Struct)
            {
                // Member is itself a struct, so call this function with that member and add all his members to the vector "memberRanges"
                std::vector<BufferRange> br = parseUniformStruct(comp, memberType, offset, memberName + ".");
                memberRanges.insert(std::end(memberRanges), std::begin(br), std::end(br));
            }
            else
            {
                // Member is not a struct. Get the size and type of it and push it to the vector.
                uint32_t range = static_cast<uint32_t>(comp.get_declared_struct_member_size(spirType, i));
                DataType dataType = getDataType(memberType);
                memberRanges.push_back({ memberName, offset, range, i, dataType });
            }

        }

        return memberRanges;
    }

// Get called like this:
for (auto &resource : resources.uniform_buffers)
{
    // Get name, offsets and ranges for individual data-members in a struct. Recursive algorithm which reads also structs in structs and so on.
    std::vector<BufferRange> memberRanges = parseUniformStruct(comp, type, 0, "");
}
```

**Comment by soconne** [comment]
You are correct, as Khronos has updated their documentation on SPIRV-Cross concerning full reflection of struct members (**Struct Member Reflection**)
https://github.com/KhronosGroup/SPIRV-Cross/wiki/Reflection-API-user-guide

I will add full support for this as well.  However returning a linear list with name strings prefixed by the parent struct does not seem ideal, in particular for cases with multiple levels of structs.  Maintaining the original hierarchy during traversal would be preferred.

**Comment by soconne** [comment]
Implemented 501ff43da2b0a472f8bb564d6290f635e7521014
